### PR TITLE
Feature/oscillator base class

### LINKED ISSUE #1: Oscillator base class
Should take care of broadcasting frequency and phase to appropriate dimensions, and computing phase from instantaneous frequency

### PR DESCRIPTION
**Pull request by ben-hayes**
Closes #1 

Base class implemented according to specs on issue. Also added forwarding of keyword args to allow for more flexible oscillator subclasses (e.g. pulse-width parameters).

### CODE REVIEW

File: test/test_oscillator.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    check_computes_correct_phase(
+        mock_dummy_osc,
+        torch.Tensor([[math.pi]]),
+        torch.Tensor([[0]]),
+        4,
+        None,
+        torch.Tensor([[0, math.pi, 2 * math.pi, 3 * math.pi]]))
+
+
+def test_phase_mod_is_optional(mock_dummy_osc):
```
Comment (vcyrus) [inline test/test_oscillator.py:198]: How about for a batch size > 1?

File: test/test_oscillator.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+def test_throws_if_length_given_when_frequency_time_axis_used(mock_dummy_osc):
+    check_for_length_mismatch_error(
+        mock_dummy_osc,
+        torch.Tensor([1, 2, 3]),
+        torch.Tensor([0]),
+        4,
+        4)
+
+
+def test_throws_if_length_given_when_phase_time_axis_used(mock_dummy_osc):
```
Comment (vcyrus) [inline test/test_oscillator.py:251]: add test_throws_if_length_given_when_freq_time_axis_used_and_phase_unspecified(...)

File: test/test_oscillator.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    check_for_length_mismatch_error(
+        mock_dummy_osc,
+        torch.Tensor([0]),
+        torch.Tensor([1, 2, 3]),
+        4,
+        4)
```
Comment (vcyrus) [inline test/test_oscillator.py:257]: what happens when length is 0 and time-axis is not used in freq/phase?

File: gradrack/oscillators/oscillator_base.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        original_frequency_length = frequency.shape[-1]
+        frequency = F.pad(frequency, (1, 0), 'constant', 0)
+        frequency = frequency.narrow(-1, 0, original_frequency_length)
+
+        return frequency
+
+    def _compute_phase(self, frequency, phase_mod, sample_rate):
+        # Integrate instantaneous frequency along the time axis to get phase
+        phase = frequency.cumsum(-1)
```
[the remark this answers is not in the document]
Comment (ben-hayes) [reply gradrack/oscillators/oscillator_base.py:109]: Yep, covers both cases. Per-sample angular frequency is achieved by setting sample_rate to math.tau:
```python
    def _replace_empty_sample_rate(self, sample_rate):
        # If sample rate is empty, set it to 2π
        return sample_rate or math.tau
 ```

In terms of when to convert to angular frequency, the next line does this:
```python
        phase = math.tau * phase / sample_rate
```

But it doesn't make any difference when it happens as it's a linear operation. I don't have strong feelings about whether to put this before or after so happy to go with whatever you prefer :)

[the remark this answers is not in the document]
Comment (vcyrus) [reply gradrack/oscillators/oscillator_base.py:109]: I see, so ultimately math.tau cancels if we are working with angular frequency

**Pull request merged**